### PR TITLE
cli_parser: Common switch options to be accessible across brew

### PR DESCRIPTION
**Pull request by GauthamGoli**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
#1860 

~As indicated above this a WIP PR, open to feedback.~

Context: Arguments like `--verbose` , `--debug`, `--quiet`, `--force` are to be accessible across commands, they cannot be stored as `Hombrew::CLI::Parser`'s instance variables.
Example: Command `A` might use a method defined elsewhere, and this method could check for `--debug` 

~To achieve this, I'm storing common `switch`es listed above in a class variable of `Homebrew::CLI::Parser` , and I've declared `verbose` and `debug` in every `dev-cmd` that uses `Homebrew::CLI::Parser`~

### CODE REVIEW

File: Library/Homebrew/cli_parser.rb [hunk 1]
```diff
@@ -4,10 +4,17 @@
 module Homebrew
   module CLI
     class Parser
+      # rubocop:disable Style/ClassVars
+      @@args = OpenStruct.new
```
Comment (GauthamGoli) [inline Library/Homebrew/cli_parser.rb]: @MikeMcQuaid Storing this in `Homebrew::CLI::Parser` class as class variable currently, as Parser is not going to be inherited but used as is. 
If we'd consider class variables evil, I'll move this to `Homebrew` class instance variable as discussed on slack.

File: Library/Homebrew/extend/ARGV.rb [hunk 2]
```diff
@@ -156,15 +158,16 @@ def values(name)
...
 
   def force?
-    flag? "--force"
+    flag?("--force") || Homebrew::CLI::Parser.args.force?
   end
 
   def verbose?
-    flag?("--verbose") || !ENV["VERBOSE"].nil? || !ENV["HOMEBREW_VERBOSE"].nil?
+    flag?("--verbose") || !ENV["VERBOSE"].nil? || !ENV["HOMEBREW_VERBOSE"].nil? ||
+      Homebrew::CLI::Parser.args.verbose?
```
Comment (GauthamGoli) [inline Library/Homebrew/extend/ARGV.rb]: Instead of adding an `or` condition (`|| Homebrew::CLI::Parser.args.verbose?`) wherever `ARGV.verbose?` has been used throughout codebase, I'm adding it here and ideally would replace occurrences `ARGV.verbose?` with `Homebrew::CLI::Parser.args.verbose?` once we use `Parser` everywhere else.

File: Library/Homebrew/cli_parser.rb [hunk 3]
```diff
@@ -4,10 +4,17 @@
...
+      # rubocop:enable Style/ClassVars
       def self.parse(&block)
         new(&block).parse
       end
 
+      def self.args
+        @@args
```
Comment (reitermarkus) [inline Library/Homebrew/cli_parser.rb]: You can use `@args ||= OpenStruct.new` here instead of the class variable.

File: Library/Homebrew/cli_parser.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cli_parser.rb]: > If we'd consider class variables evil

Yup 👍 

> I'll move this to Homebrew class instance variable as discussed on slack.

To clarify to other readers here: I suggested having `Homebrew.args` as a global instance kinda like `Homebrew.failed` currently is.

File: Library/Homebrew/cli_parser.rb (hunk 3 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cli_parser.rb]: This is another option but I do think a global `Homebrew.args` is nice and snappy.

File: Library/Homebrew/cli_parser.rb [hunk 4]
```diff
@@ -16,11 +23,13 @@ def initialize(&block)
...
         description = option_to_description(*names) if description.nil?
-        names, env = common_switch(*names) if names.first.is_a?(Symbol)
+        global_switch = names.first.is_a?(Symbol)
+        names, env = common_switch(*names) if global_switch
         @parser.on(*names, description) do
-          enable_switch(*names)
+          enable_switch(*names, global_switch)
         end
-        enable_switch(*names) if !env.nil? && !ENV["HOMEBREW_#{env.to_s.upcase}"].nil?
+        enable_switch(*names, global_switch) if !env.nil? &&
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli_parser.rb]: I'd use an `if ... end` here.

File: Library/Homebrew/cli_parser.rb [hunk 5]
```diff
@@ -57,17 +66,20 @@ def parse(cmdline_args = ARGV)
 
       private
 
-      def enable_switch(*names)
+      def enable_switch(*names, global_switch)
         names.each do |name|
+          next @@args["#{option_to_name(name)}?"] = true if global_switch
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli_parser.rb]: I'd use an `if .. end` here and split out the `next` from assignment.

File: Library/Homebrew/dev-cmd/edit.rb [hunk 6]
```diff
@@ -41,7 +43,7 @@ def edit
       paths = ARGV.named.map do |name|
         path = Formulary.path(name)
 
-        raise FormulaUnavailableError, name unless path.file? || args.force?
+        raise FormulaUnavailableError, name unless path.file? || ARGV.force?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/edit.rb]: Why this change?

File: Library/Homebrew/extend/ARGV.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/extend/ARGV.rb]: I think it would be good to not add any references to `Homebrew::CLI::Parser.args` or `Homebrew.args` in `ARGV` but instead work towards replacing calls to `ARGV` with calls to `Homebrew.args` instead.

File: Library/Homebrew/cli_parser.rb (hunk 4 above)
Comment (GauthamGoli) [reply Library/Homebrew/cli_parser.rb]: This leads to Rubocop error  : `Use a guard clause instead of wrapping the code inside a conditional expression.`

**Comment by GauthamGoli** [comment]
@MikeMcQuaid Addressed comments and updated the PR

**Comment by GauthamGoli** [comment]
@MikeMcQuaid I've updated `bottle.rb` to use `Homebrew.args` for global options

**Comment by MikeMcQuaid** [comment]
Thanks again @GauthamGoli, great work as usual!
